File: src/presentation/camera.ts
/**
 * 2.5D projection.
 *
 * The rules layer stores a flat grid plus an `elevation` per tile. This module
 * turns (tileX, tileY, elevation) into screen coordinates and a depth key.
 * It is pure maths with no canvas or engine types, so a Godot or Unity port
 * reuses it verbatim.
 *
 * The projection is dimetric: the world is rotated 45 degrees and squashed
 * vertically, which is the classic isometric pixel-art reading, then elevation
 * is subtracted from the screen Y. That last part is the whole trick - height
 * in the world is simply "further up the screen", and the depth key keeps the
 * draw order correct.
 */

export interface Vec3 {
  x: number;
  y: number;
  /** Elevation in tile units. */
  z: number;
}

export interface ScreenPoint {
  x: number;
  y: number;
  /** Painter's-algorithm sort key. Higher draws later (in front). */
  depth: number;
}

export interface CameraConfig {
  /** Half-width of a tile in screen pixels. */
  tileHalfWidth: number;
  /** Half-height of a tile in screen pixels. Lower means a flatter angle. */
  tileHalfHeight: number;
  /** Screen pixels one unit of elevation raises a point. */
  elevationUnit: number;
  /** Camera focus in world tile coordinates. */
  focus: Vec3;
  /** Viewport size in pixels. */
  viewportWidth: number;
  viewportHeight: number;
  /** Rendering scale. Keep this an integer to preserve crisp pixels. */
  zoom: number;
  /**
   * Vertical framing offset in screen pixels, applied after projection.
   *
   * Sprites are drawn upward from their ground point, so a scene centred on
   * the ground plane sits high in the frame. A positive offset pushes the
   * world down to rebalance it, and it is also how you leave room for a UI
   * bar without moving the camera focus off the action.
   */
  offsetY?: number;
}

export function defaultCamera(viewportWidth: number, viewportHeight: number): CameraConfig {
  return {
    tileHalfWidth: 32,
    tileHalfHeight: 16,
    elevationUnit: 20,
    focus: { x: 0, y: 0, z: 0 },
    viewportWidth,
    viewportHeight,
    zoom: 1,
    offsetY: 0,
  };
}

/**
 * World tile coordinates to screen pixels.
 *
 * Depth is computed from the world position rather than the screen position so
 * that a tall object on a low tile still sorts behind a short object on a high
 * tile when it should.
 */
export function project(world: Vec3, camera: CameraConfig): ScreenPoint {
  const dx = world.x - camera.focus.x;
  const dy = world.y - camera.focus.y;
  const dz = world.z - camera.focus.z;

  const screenX = (dx - dy) * camera.tileHalfWidth * camera.zoom;
  const screenY = ((dx + dy) * camera.tileHalfHeight - dz * camera.elevationUnit) * camera.zoom;

  return {
    x: screenX + camera.viewportWidth / 2,
    y: screenY + camera.viewportHeight / 2 + (camera.offsetY ?? 0),
    // Ties on (x+y) break by elevation, so higher objects draw in front.
    depth: (world.x + world.y) * 16 + world.z,
  };
}

/** Screen pixels back to world tile coordinates on a given elevation plane. */
export function unproject(screenX: number, screenY: number, camera: CameraConfig, planeZ = 0): Vec3 {
  const localX = (screenX - camera.viewportWidth / 2) / camera.zoom;
  const localY =
    (screenY - camera.viewportHeight / 2 - (camera.offsetY ?? 0)) / camera.zoom + planeZ * camera.elevationUnit;

  const a = localX / camera.tileHalfWidth;
  const b = localY / camera.tileHalfHeight;

  return {
    x: (b + a) / 2 + camera.focus.x,
    y: (b - a) / 2 + camera.focus.y,
    z: planeZ,
  };
}

/**
 * Snap a screen point to whole pixels.
 *
 * Sub-pixel sprite positions are the cause of the shimmering that ruins
 * pixel art under a moving camera. Everything drawn should pass through here.
 */
export function snapToPixel(point: ScreenPoint): ScreenPoint {
  return { x: Math.round(point.x), y: Math.round(point.y), depth: point.depth };
}

/** Move the camera toward a target, framerate-independent. */
export function followCamera(camera: CameraConfig, target: Vec3, smoothing: number, dtMs: number): CameraConfig {
  // Exponential smoothing expressed so the result is independent of framerate.
  const factor = 1 - Math.pow(1 - smoothing, dtMs / 16.667);
  return {
    ...camera,
    focus: {
      x: camera.focus.x + (target.x - camera.focus.x) * factor,
      y: camera.focus.y + (target.y - camera.focus.y) * factor,
      z: camera.focus.z + (target.z - camera.focus.z) * factor,
    },
  };
}

/** Whether a projected point is inside the viewport, with a margin. */
export function isVisible(point: ScreenPoint, camera: CameraConfig, margin = 64): boolean {
  return (
    point.x >= -margin &&
    point.y >= -margin &&
    point.x <= camera.viewportWidth + margin &&
    point.y <= camera.viewportHeight + margin
  );
}

File: unity/Aetherlight.Core/Presentation/Camera.cs
using System;

namespace Aetherlight.Presentation
{
    public struct Vec3
    {
        public double X;
        public double Y;
        /// <summary>Elevation in tile units.</summary>
        public double Z;

        public Vec3(double x, double y, double z = 0)
        {
            X = x;
            Y = y;
            Z = z;
        }

        public override string ToString() => $"({X}, {Y}, {Z})";
    }

    public struct ScreenPoint
    {
        public double X;
        public double Y;
        /// <summary>Painter's-algorithm sort key. Higher draws later (in front).</summary>
        public double Depth;
    }

    public sealed class CameraConfig
    {
        /// <summary>Half-width of a tile in screen pixels.</summary>
        public double TileHalfWidth = 32;

        /// <summary>Half-height of a tile. Lower means a flatter angle.</summary>
        public double TileHalfHeight = 16;

        /// <summary>Screen pixels one unit of elevation raises a point.</summary>
        public double ElevationUnit = 20;

        /// <summary>Camera focus in world tile coordinates.</summary>
        public Vec3 Focus;

        public double ViewportWidth;
        public double ViewportHeight;

        /// <summary>Rendering scale. Keep this an integer to preserve crisp pixels.</summary>
        public double Zoom = 1;

        /// <summary>
        /// Vertical framing offset in screen pixels, applied after projection.
        ///
        /// Sprites are drawn upward from their ground point, so a scene centred
        /// on the ground plane sits high in the frame. A positive offset pushes
        /// the world down to rebalance it, and it is also how you leave room for
        /// a UI bar without moving the focus off the action.
        /// </summary>
        public double OffsetY;

        public static CameraConfig Default(double viewportWidth, double viewportHeight) => new CameraConfig
        {
            ViewportWidth = viewportWidth,
            ViewportHeight = viewportHeight,
        };

        public CameraConfig Clone() => new CameraConfig
        {
            TileHalfWidth = TileHalfWidth,
            TileHalfHeight = TileHalfHeight,
            ElevationUnit = ElevationUnit,
            Focus = Focus,
            ViewportWidth = ViewportWidth,
            ViewportHeight = ViewportHeight,
            Zoom = Zoom,
            OffsetY = OffsetY,
        };
    }

    /// <summary>
    /// 2.5D projection.
    ///
    /// The rules layer stores a flat grid plus an elevation per tile. This
    /// turns (x, y, z) into screen coordinates and a depth key. Pure maths with
    /// no engine types, so a Unity renderer reuses it verbatim.
    ///
    /// The projection is dimetric: the world is rotated 45 degrees and squashed
    /// vertically, then elevation is subtracted from screen Y. That last part is
    /// the whole trick - height in the world is simply "further up the screen",
    /// and the depth key keeps the draw order correct.
    /// </summary>
    public static class Camera
    {
        public static ScreenPoint Project(Vec3 world, CameraConfig camera)
        {
            double dx = world.X - camera.Focus.X;
            double dy = world.Y - camera.Focus.Y;
            double dz = world.Z - camera.Focus.Z;

            double screenX = (dx - dy) * camera.TileHalfWidth * camera.Zoom;
            double screenY = ((dx + dy) * camera.TileHalfHeight - dz * camera.ElevationUnit) * camera.Zoom;

            return new ScreenPoint
            {
                X = screenX + camera.ViewportWidth / 2,
                Y = screenY + camera.ViewportHeight / 2 + camera.OffsetY,
                // Ties on (x + y) break by elevation, so higher objects draw in front.
                Depth = (world.X + world.Y) * 16 + world.Z,
            };
        }

        /// <summary>Screen pixels back to world tile coordinates on an elevation plane.</summary>
        public static Vec3 Unproject(double screenX, double screenY, CameraConfig camera, double planeZ = 0)
        {
            double localX = (screenX - camera.ViewportWidth / 2) / camera.Zoom;
            double localY = (screenY - camera.ViewportHeight / 2 - camera.OffsetY) / camera.Zoom
                            + planeZ * camera.ElevationUnit;

            double a = localX / camera.TileHalfWidth;
            double b = localY / camera.TileHalfHeight;

            return new Vec3((b + a) / 2 + camera.Focus.X, (b - a) / 2 + camera.Focus.Y, planeZ);
        }

        /// <summary>
        /// Snap to whole pixels. Sub-pixel sprite positions are what make pixel
        /// art shimmer under a moving camera, so everything drawn passes here.
        /// </summary>
        public static ScreenPoint SnapToPixel(ScreenPoint point) => new ScreenPoint
        {
            X = Math.Round(point.X, MidpointRounding.AwayFromZero),
            Y = Math.Round(point.Y, MidpointRounding.AwayFromZero),
            Depth = point.Depth,
        };

        /// <summary>Move the camera toward a target, framerate-independent.</summary>
        public static CameraConfig Follow(CameraConfig camera, Vec3 target, double smoothing, double dtMs)
        {
            // Exponential smoothing expressed so the result does not depend on
            // how often it is called.
            double factor = 1 - Math.Pow(1 - smoothing, dtMs / 16.667);
            var moved = camera.Clone();
            moved.Focus = new Vec3(
                camera.Focus.X + (target.X - camera.Focus.X) * factor,
                camera.Focus.Y + (target.Y - camera.Focus.Y) * factor,
                camera.Focus.Z + (target.Z - camera.Focus.Z) * factor);
            return moved;
        }

        public static bool IsVisible(ScreenPoint point, CameraConfig camera, double margin = 64) =>
            point.X >= -margin
            && point.Y >= -margin
            && point.X <= camera.ViewportWidth + margin
            && point.Y <= camera.ViewportHeight + margin;
    }
}
